[tests] Introduce TestNode

jnewbery opened the pull request:
Continues #10082 

TestNode is a class responsible for all state related to a bitcoind node
under test. It stores local state, is responsible for tracking the
bitcoind process and delegates unrecognised messages to the RPC
connection.

This commit changes start_nodes and stop_nodes to start and stop the
bitcoind nodes in parallel, making test setup and teardown much faster.


On my vm, this changeset reduces total test_runner runtime for the base set of tests
(including building the cache) from 263s to 195s (a 25% speedup). Note that the time
reported by test_runner does not include time spent building the cache:

*with TestNode*:
```
→ date +"%T" ; ./test_runner.py -q ; date +"%T"
12:48:04
..................................................................................................................................................................................................................................................................................................................................
TEST                           | STATUS    | DURATION

abandonconflict.py             | ✓ Passed  | 12 s
bip68-112-113-p2p.py           | ✓ Passed  | 19 s
blockchain.py                  | ✓ Passed  | 8 s 
bumpfee.py                     | ✓ Passed  | 13 s
decodescript.py                | ✓ Passed  | 3 s 
disablewallet.py               | ✓ Passed  | 3 s 
disconnect_ban.py              | ✓ Passed  | 6 s 
fundrawtransaction.py          | ✓ Passed  | 37 s
getchaintips.py                | ✓ Passed  | 4 s 
httpbasics.py                  | ✓ Passed  | 3 s 
import-rescan.py               | ✓ Passed  | 4 s 
importmulti.py                 | ✓ Passed  | 6 s 
importprunedfunds.py           | ✓ Passed  | 3 s 
invalidblockrequest.py         | ✓ Passed  | 4 s 
invalidtxrequest.py            | ✓ Passed  | 4 s 
keypool.py                     | ✓ Passed  | 7 s 
listsinceblock.py              | ✓ Passed  | 4 s 
listtransactions.py            | ✓ Passed  | 33 s
mempool_limit.py               | ✓ Passed  | 4 s 
mempool_persist.py             | ✓ Passed  | 15 s
mempool_reorg.py               | ✓ Passed  | 4 s 
mempool_resurrect_test.py      | ✓ Passed  | 3 s 
mempool_spendcoinbase.py       | ✓ Passed  | 3 s 
merkle_blocks.py               | ✓ Passed  | 3 s 
multi_rpc.py                   | ✓ Passed  | 4 s 
net.py                         | ✓ Passed  | 3 s 
nulldummy.py                   | ✓ Passed  | 3 s 
p2p-compactblocks.py           | ✓ Passed  | 28 s
p2p-fullblocktest.py           | ✓ Passed  | 126 s
p2p-leaktests.py               | ✓ Passed  | 8 s 
p2p-mempool.py                 | ✓ Passed  | 3 s 
p2p-segwit.py                  | ✓ Passed  | 59 s
p2p-versionbits-warning.py     | ✓ Passed  | 8 s 
preciousblock.py               | ✓ Passed  | 3 s 
prioritise_transaction.py      | ✓ Passed  | 5 s 
proxy_test.py                  | ✓ Passed  | 3 s 
rawtransactions.py             | ✓ Passed  | 9 s 
receivedby.py                  | ✓ Passed  | 19 s
reindex.py                     | ✓ Passed  | 12 s
rest.py                        | ✓ Passed  | 9 s 
rpcnamedargs.py                | ✓ Passed  | 3 s 
segwit.py                      | ✓ Passed  | 7 s 
sendheaders.py                 | ✓ Passed  | 24 s
signmessages.py                | ✓ Passed  | 3 s 
signrawtransactions.py         | ✓ Passed  | 3 s 
txn_clone.py                   | ✓ Passed  | 4 s 
txn_doublespend.py --mineblock | ✓ Passed  | 4 s 
uptime.py                      | ✓ Passed  | 3 s 
wallet-accounts.py             | ✓ Passed  | 3 s 
wallet-dump.py                 | ✓ Passed  | 7 s 
wallet-encryption.py           | ✓ Passed  | 8 s 
wallet-hd.py                   | ✓ Passed  | 15 s
wallet.py                      | ✓ Passed  | 31 s
walletbackup.py                | ✓ Passed  | 104 s
zapwallettxes.py               | ✓ Passed  | 9 s 
zmq_test.py                    | ○ Skipped | 0 s 

ALL                            | ✓ Passed  | 735 s (accumulated)
Runtime: 189 s

12:51:19
```

*master*:
```
→ date +"%T" ; ./test_runner.py -q ; date +"%T"
12:40:13
..........................................................................................................................................................................................................................................................................................................................................................................................................................................
TEST                           | STATUS    | DURATION

abandonconflict.py             | ✓ Passed  | 15 s
bip68-112-113-p2p.py           | ✓ Passed  | 19 s
blockchain.py                  | ✓ Passed  | 8 s
bumpfee.py                     | ✓ Passed  | 20 s
decodescript.py                | ✓ Passed  | 3 s
disablewallet.py               | ✓ Passed  | 3 s
disconnect_ban.py              | ✓ Passed  | 8 s
fundrawtransaction.py          | ✓ Passed  | 36 s
getchaintips.py                | ✓ Passed  | 11 s
httpbasics.py                  | ✓ Passed  | 7 s
import-rescan.py               | ✓ Passed  | 16 s
importmulti.py                 | ✓ Passed  | 10 s
importprunedfunds.py           | ✓ Passed  | 5 s
invalidblockrequest.py         | ✓ Passed  | 4 s
invalidtxrequest.py            | ✓ Passed  | 3 s
keypool.py                     | ✓ Passed  | 7 s
listsinceblock.py              | ✓ Passed  | 11 s
listtransactions.py            | ✓ Passed  | 37 s
mempool_limit.py               | ✓ Passed  | 4 s
mempool_persist.py             | ✓ Passed  | 23 s
mempool_reorg.py               | ✓ Passed  | 7 s
mempool_resurrect_test.py      | ✓ Passed  | 3 s
mempool_spendcoinbase.py       | ✓ Passed  | 3 s
merkle_blocks.py               | ✓ Passed  | 10 s
multi_rpc.py                   | ✓ Passed  | 6 s
net.py                         | ✓ Passed  | 6 s
nulldummy.py                   | ✓ Passed  | 3 s
p2p-compactblocks.py           | ✓ Passed  | 30 s
p2p-fullblocktest.py           | ✓ Passed  | 126 s
p2p-leaktests.py               | ✓ Passed  | 8 s
p2p-mempool.py                 | ✓ Passed  | 3 s
p2p-segwit.py                  | ✓ Passed  | 62 s
p2p-versionbits-warning.py     | ✓ Passed  | 8 s
preciousblock.py               | ✓ Passed  | 8 s
prioritise_transaction.py      | ✓ Passed  | 7 s
proxy_test.py                  | ✓ Passed  | 10 s
rawtransactions.py             | ✓ Passed  | 15 s
receivedby.py                  | ✓ Passed  | 28 s
reindex.py                     | ✓ Passed  | 12 s
rest.py                        | ✓ Passed  | 12 s
rpcnamedargs.py                | ✓ Passed  | 3 s
segwit.py                      | ✓ Passed  | 12 s
sendheaders.py                 | ✓ Passed  | 26 s
signmessages.py                | ✓ Passed  | 3 s
signrawtransactions.py         | ✓ Passed  | 3 s
txn_clone.py                   | ✓ Passed  | 10 s
txn_doublespend.py --mineblock | ✓ Passed  | 10 s
uptime.py                      | ✓ Passed  | 3 s
wallet-accounts.py             | ✓ Passed  | 3 s
wallet-dump.py                 | ✓ Passed  | 6 s
wallet-encryption.py           | ✓ Passed  | 8 s
wallet-hd.py                   | ✓ Passed  | 18 s
wallet.py                      | ✓ Passed  | 69 s
walletbackup.py                | ✓ Passed  | 130 s
zapwallettxes.py               | ✓ Passed  | 15 s
zmq_test.py                    | ○ Skipped | 0 s

ALL                            | ✓ Passed  | 936 s (accumulated)
Runtime: 242 s

12:44:36
```

jnewbery committed:
[tests] Introduce TestNode

TestNode is a class responsible for all state related to a bitcoind node
under test. It stores local state, is responsible for tracking the
bitcoind process and delegates unrecognised messages to the RPC
connection.

This commit changes start_nodes and stop_nodes to start and stop the
bitcoind nodes in parallel, making test setup and teardown much faster.

jnewbery commented:
Wallet tests were failing on Travis because bitcoind startup was timing out. Increasing timeout from 2.5s to 10s for Travis.

laanwj commented:
The speed-up is nice, also like the idea of launching tests in parallel. Concept ACK.

jtimon commented on test/functional/test_framework/util.py:
```diff
@@ -244,6 +244,7 @@ def get_auth_cookie(datadir, n):
                 if line.startswith("rpcpassword="):
                     assert password is None  # Ensure that there is only one rpcpassword line
                     password = line.split("=")[1].strip("\n")
+    # print(os.path.join(datadir, "regtest", ".cookie"))
```
Remove this?

jnewbery replied:
thanks. Removed

jnewbery commented:
fixed @jtimon's review comment and rebased.

jnewbery commented:
rebased with TestNode changes for multiwallet.

MarcoFalke commented:
Needs rebase

jnewbery commented:
rebased

@MarcoFalke @kallewoof @jimmysong @NicolasDorier I'd love some review of this if any of you have time.

jimmysong commented on test/functional/test_framework/test_node.py:
```diff
@@ -0,0 +1,130 @@
+        """Sets up an RPC connection to the bitcoind process. Returns False if unable to connect."""
+
+        attempts = 40
+        while attempts > 0:
```
This is a little harder to read than something like this:

    for attempt in range(40):
        ...
        time.sleep(0.25)

jnewbery replied:
Thanks. changed

jimmysong commented on test/functional/test_framework/test_node.py:
```diff
@@ -0,0 +1,130 @@
+        self.log.debug("Stopping node")
+        try:
+            self.stop()
+        except http.client.CannotSendRequest as e:
```
no need for `as e`

jnewbery replied:
Yep

jimmysong commented on test/functional/test_framework/test_node.py:
```diff
@@ -91,6 +91,12 @@ def wait_for_rpc_connection(self):
+        assert self.rpc_connected
+        assert self.rpc
+        wallet_path = "wallet/%s" % wallet_name
+        return self.rpc / wallet_path
```
What does this do? `wallet_path` is for sure a string, but how do you divide the rpc object by a string?

jnewbery replied:
By magic :)

Take a look at https://github.com/bitcoin/bitcoin/blob/e526ca6284b9e13be1b912b80dd73a34e739b539/test/functional/test_framework/authproxy.py#L195 and https://github.com/bitcoin/bitcoin/blob/e526ca6284b9e13be1b912b80dd73a34e739b539/test/functional/test_framework/coverage.py#L59

The interface here is supposed to be similar to Python's [pathlib](https://docs.python.org/3/library/pathlib.html)

This functionality was added to support the multiwallet endpoints in https://github.com/bitcoin/bitcoin/commit/979d0b8a6533de58b6352666d90fff33841db63d . Take a look at the `multiwallet.py` test to see how it's currently used.

jimmysong reviewed (commented):
utACK. Couple of nits and one thing I don't get.

NicolasDorier commented on test/functional/test_framework/test_framework.py:
```diff
@@ -203,67 +197,60 @@ def main(self):
-        return proxy
+        return node
 
     def start_nodes(self, num_nodes, dirname, extra_args=None, rpchost=None, timewait=None, binary=None):
```
maybe default to `extra_args=[]` to be coherent with above ?

NicolasDorier replied:
Ah actually no, this is a array of array....

NicolasDorier commented:
utACK 2d189c89f6ecdb846a046bdd57883e190df7b3bf I am a big fan of this idea, it makes writing tests so much easier.

kallewoof commented on test/functional/test_framework/test_framework.py:
```diff
@@ -203,67 +197,60 @@ def main(self):
     # Public helper methods. These can be accessed by the subclass test scripts.
 
-    def start_node(self, i, dirname, extra_args=None, rpchost=None, timewait=None, binary=None, stderr=None):
+    def start_node(self, i, dirname, extra_args=[], rpchost=None, timewait=None, binary=None, stderr=None):
```
Why not allow `None` for `extra_args`? I'm generally wary of passing in objects as defaults, as if those default objects are ever altered, the altered version will be passed as the default in future calls.

E.g. try the following snippet:
```Python
def foo(x=[]):
    x.append(5)
    print(x)

foo()
foo()
foo()
```
You will get as output
```
[5]
[5, 5]
[5, 5, 5]
```

NicolasDorier replied:
wow I did not know.

In general I think that the user should be able to pass `null` (Nothing), when he means "default value".

This code can be easily fixed by keeping the default value to Nothing and first line, checking if extra_args is nothing, if that is the case, init to an empty array.

jimmysong replied:
Totally didn't catch that, but I've run into this in the past and it's pretty nasty to debug unless you know to look for it. I agree with the others, put this back to None.

kallewoof replied:
`null` in Python is `None`, so yeah, I think `None` should be the default here and just set it to `[]` if it's `None` at the start.

jnewbery replied:
Nice. Thanks @kallewoof , I wasn't aware of that, but it makes perfect sense.

Changed the default argument to `None` as suggested.

kallewoof reviewed (approved):
utACK with one nit.

MarcoFalke commented on test/functional/test_framework/test_node.py:
```diff
@@ -0,0 +1,135 @@
+
+    def start(self):
+        """Start the node."""
+        # import pdb; pdb.set_trace()
```
nit: debug artifact

jnewbery replied:
removed

MarcoFalke commented on test/functional/test_framework/test_framework.py:
```diff
@@ -206,64 +200,59 @@ def main(self):
-        return rpcs
+
+        if self.options.coveragedir is not None:
+            coverage.write_all_rpc_commands(self.options.coveragedir, node.rpc)
```
missing for node in nodes?

jnewbery replied:
fixed

MarcoFalke commented on test/functional/test_framework/test_framework.py:
```diff
@@ -206,64 +200,59 @@ def main(self):
-        assert not self.bitcoind_processes.values()  # All connections must be gone now
+        for node in self.nodes:
+            node.stop_node()
+        # All connections must be gone now
```
nit: remove outdated comment

jnewbery replied:
removed

MarcoFalke commented on test/functional/test_framework/test_node.py:
```diff
@@ -0,0 +1,135 @@
+                self.rpc_connected = True
+                self.url = self.rpc.url
+                self.log.debug("RPC successfully started")
+                return True
```
nit: No need to return a value if it is never used. Can be replaced by a plain `return`.

jnewbery replied:
fixed

MarcoFalke commented on test/functional/test_framework/test_framework.py:
```diff
@@ -206,64 +200,59 @@ def main(self):
         try:
             for i in range(num_nodes):
-                rpcs.append(self.start_node(i, dirname, extra_args[i], rpchost, timewait=timewait, binary=binary[i]))
+                nodes.append(TestNode(i, dirname, extra_args[i], rpchost, timewait=timewait, binary=binary[i], stderr=None, mocktime=self.mocktime, coverage_dir=self.options.coveragedir))
```
stderr must be set to not None. Otherwise, we will ignore any output to stderr, no?

jnewbery replied:
This is the same as the existing behaviour. `start_nodes()` currently calls `start_node()` without the `stderr` argument set, so it defaults to `None`.

This could be changed in a future PR.

maflcko reviewed (changes requested):
utACK cb8579bd3fb96d6204c07bf3b2d2476d0f077a91, but needs some nits and apparent regressions fixed.

jnewbery commented on test/functional/multiwallet.py:
```diff
@@ -35,11 +35,15 @@ def run_test(self):
+        w1 = self.nodes[0].get_wallet_rpc("w1")
+        w2 = self.nodes[0].get_wallet_rpc("w2")
+        w3 = self.nodes[0].get_wallet_rpc("w3")
+        wallet_bad = self.nodes[0].get_wallet_rpc("wallet/bad")
```
fixed

jnewbery commented:
Thanks for the review @MarcoFalke - I've addressed all your nits (except the `stderr` in `start_nodes()` which I think maintains existing behaviour).

I've updated and squashed the commits. Unsquashed history is here: https://github.com/jnewbery/bitcoin/tree/pr10711.1 if you just want to see what's changed.

MarcoFalke commented:
re-utACK 7897338, thanks for addressing the feedback.

I keep a local copy of all commits, so no need to provide branches of the history for me.
